Remove theano.dot

eigenfoo opened the pull request:
~~[WIP, will request review when ready.]~~

Closes #52.

I've discovered what `theano.dot` does - it's there to provide a sparse-agnostic way of multiplying two tensors. I think this PR is well summarized by this table, where the bolded methods implement the "if-failed-then-try-`rdot`" logic.

|            | Regular                                       | Sparse       |
|------------|-----------------------------------------------|--------------|
| Previously | **`theano.dot`**, `theano.tensor.dot`, `x.dot(y)` | **`theano.dot`** |
| If merged  | `theano.tensor.dot`, **`x.dot(y)`**             | **`x.dot(y)`**   |

Previously, a user could just use `theano.dot` for the highest-level call. Now, it will be `x.dot(y)`. Functional-minded users might not like this - i.e. some users might prefer calling `foo(x, y)` instead of `x.foo(y)`. More importantly, if this PR is merged, the only way to multiply sparse tensors will unintuitively be with the `.dot()` method, instead of any top-level function.

What do we think?

brandonwillard commented:
If we move the functionality of the old `theano.dot` into `theano.tensor.dot` then we could have both, no?

eigenfoo commented:
Yup, no reason why not. I was a bit hesitant to mix sparse and dense functionality into `theano.tensor`, but eventually we want to restructure how sparse tensors work anyways (#142), so this will be a step in the right direction. Let me fix that.

eigenfoo commented:
To avoid fairly ugly recursion errors, I've just moved the implementation of `theano.dot` into `theano.tensor` module, and renamed the existing `theano.tensor.dot` to `theano.tensor.dense_dot`, to be more explicit.

To recap, here's the table again, with the fixes:

|            | Regular                                       | Sparse       |
|------------|-----------------------------------------------|--------------|
| Previously | **`theano.dot`**, `theano.tensor.dot`, `x.dot(y)` | **`theano.dot`** |
| If merged  | **`theano.tensor.dot`**, `theano.tensor.dense_dot(x, y)`, `x.dot(y)`             | **`theano.tensor.dot`**, `x.dot(y)`   |

brandonwillard commented on theano/tensor/basic.py:
```diff
@@ -6306,7 +6306,22 @@ def __str__(self):
+    This is designed to work with both sparse and dense tensors types.
+    """
+    try:
+        res = l.__dot__(r)
```
What if `l.__dot__` _returns_ `NotImplemented`; wouldn't we want to try `r.__rdot__`?

eigenfoo replied:
Hmm, that's not what the old implementation did - if `l.__dot__(r)` returned `NotImplemented`, then it would just run a `NotImplementedError`. Is that behavior we'd want? I can't imagine a situation where one returned `NotImplemented` while the other didn't.

brandonwillard replied:
The old `theano.dot` would check for a return value of `NotImplemented` and raise its own `NotImplementedError` so that the `r.__rdot__` call in the `except` would be tried (see [this statement](https://github.com/pymc-devs/Theano-PyMC/blob/master/theano/__init__.py#L171)).

eigenfoo replied:
Ah whoops sorry, don't know how I missed that. The next commit should fix this.
